testsuite/tests/Text/RDF/RDF4H/TurtleSerializerTest.hs

```diff
@@ -66,4 +71,61 @@ tests = testGroup "Turtle serializer tests"
           actual <- BS.hGetContents h
           expected @=? actual)
     ]
+
+  , testGroup "QuickCheck Tests"
+    [ testProperty "Serialized graph should have only one instance of each subject" (ioProperty . (prop_SingleSubject :: RDF TList -> IO Bool))
+    ]
   ]
+
+-- |This property ensures that a Turtle file generated from a graph will only
+-- have a single instance of a given subject as the serialzer is supposed to
+-- group all triples with a given subject.
+prop_SingleSubject :: (Rdf rdf) => RDF rdf -> IO Bool
+prop_SingleSubject g = withSystemTempFile "rdf4h-"
+                       (\_ h -> do
+                           hWriteRdf serializer h g
+                           hSeek h AbsoluteSeek 0
+                           contents <- BS.hGetContents h
+                           pure $ assertSingleSubjects contents)
+  where mappings = PrefixMappings $ Map.fromList [ ("schema", "http://schema.org/")
+                                                 , ("rdf", "http://www.w3.org/1999/02/22-rdf-syntax-ns#")
+                                                 , ("dc", "http://purl.org/dc/elements/1.1/")
+                                                 ]
+        serializer = TurtleSerializer Nothing mappings
+
+        -- Convert a Node to a string if it is a UNode or a BNode. This is
+        -- acceptable since the arbitrary instance for Node's generate only
+        -- those constructors for subjects and that's what this function is
+        -- being used on.
+        toUriString :: Node -> Maybe String
+        toUriString (UNode uriText) = Just $ T.unpack uriText
+        toUriString (BNode bid) = Just $ T.unpack bid
+        toUriString _ = Nothing
+
+        -- Convert the subjects for the given graph, g, to Strings.
+        subjects :: [String]
+        subjects = nub $ sort $ catMaybes $ toUriString <$> subjectOf <$> triplesOf g
+
+        -- Count the number of times the first ByteString is found in the second
+        -- ByteString.
+        countMatches :: ByteString  -- ^The prefix
+                     -> ByteString  -- ^The ByteString to search
+                     -> Int
+        countMatches needle haystack = countMatches' 0 needle haystack
+          where countMatches' :: Int -> ByteString -> ByteString -> Int
+                countMatches' n needle' haystack' =
+                  case BS.breakSubstring needle' haystack' of
+                    (_, r) | BS.null r -> n
+                           | otherwise -> countMatches' (n+1) needle' (BS.drop (BS.length needle') r)
+
+        -- Assert that the graph serialization (first parameter) contains
+        -- exactly one instance of the given subject (second parameter).
```

Comment:
@tmciver that seems to be subtly different from what we discussed in https://github.com/robstewart57/rdf4h/pull/99#issuecomment-673976258

Your `countMatches` function looks like the needle (the subject) must not appear anywhere in the bytestring haystack (all subject/predicate/object data).

That would mean the `prop_SingleSubject` property would not allow:

    <http://www.example.com/a> <http://www.example.com/b> <http://www.example.com/c>
    <http://www.example.com/d> <http://www.example.com/e> <http://www.example.com/a>
    <http://www.example.com/f> <http://www.example.com/g> <http://www.example.com/a>

Since, when calling `assertSingleSubject` with the subject `<http://www.example.com/a>` as its second argument, the needle `<http://www.example.com/a>` is part of the haystack in the 2nd and 3rd triples.

Have I misunderstood the intention here? Or indeed, your code?

Reply:
@robstewart57 You did not misunderstand the code: you're right, this test does not account for a subject node in object position.  Good catch!  Considering that, this property seems much more difficult to test than I thought.

As I think about this, one possible solution might be to use the Turtle parser on the input graph, count the number of times a subject node occurs in non-subject position and then add 1.  That should give the number of times that URI should occur in the ttl file.

One issue that this approach suffers from is if that URI appears multiple times in predicate position and then predicates are grouped, then that calculation will be wrong.  I think this situation is rare but I think it's possible.

Another option would be to create a special-purpose parser that counts the number of times a given URI appears in subject position.  But this seems like a lot of work and could therefore have its own bugs.

Thoughts?

Reply:
@tmciver Hi,

An alternative option is to implement an `Arbitrary` instance of a special kind of RDF graph whereby:

1. A URI that appears in a subject position may appear in multiple subject positions.
2. A URI that appears in a subject position must not appear in any predicate or object positions in the graph.

The QuickCheck API should permit these constraints, and the `prop_SingleSubject` property would then be valid.

We could do with other property tests which tests for various identity functions, e.g. https://github.com/robstewart57/rdf4h/pull/99#issuecomment-673981058 and parsing, serialising then parsing again the same RDF data content for each of the given RDF formats that we so far support.

Reply:
Rob,
A great idea!  I've been pretty busy of late but will try to work on that this week.